## Trabalho Prático do Módulo 3

In [1]:
import pandas as pd
import numpy as np
import time

#split
from sklearn.model_selection import train_test_split

#Técnicas
from sklearn.linear_model import LinearRegression
from sklearn.svm import SVC
from sklearn.cluster import KMeans

#Métricas
from sklearn import metrics
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
from sklearn.metrics import silhouette_score, davies_bouldin_score, mutual_info_score,adjusted_mutual_info_score

#visualização
from plotly.offline import iplot
import plotly.graph_objects as go
import cufflinks as cf

##### Carregando base

In [2]:
blood = pd.read_csv('bloodtransf.csv')
diabetes = pd.read_csv('diabetes_numeric.csv')
wine = pd.read_csv('wine.csv')

print("Shape blood:", blood.shape)
print("Shape diabetes:", diabetes.shape)
print("Shape wine:", wine.shape)

Shape blood: (748, 5)
Shape diabetes: (43, 3)
Shape wine: (178, 14)


### 1 - Sobre o número de atributos da base de regressão, marque a alternativa CORRETA:

In [8]:
print(diabetes.shape)
diabetes.head()

(43, 3)


age  deficit  c_peptide
0   5.2     -8.1        4.8
1   8.8    -16.1        4.1
2  10.5     -0.9        5.2
3  10.6     -7.8        5.5
4  10.4    -29.0        5.0

######  <p style="background-color:green;color:white;font-size:1.3em"> R1: Temos 3 atributos </p> 



### 2 - Sobre o número de atributos da base de classificação, marque a alternativa CORRETA:

In [9]:
print(blood.shape)
blood.head()

(748, 5)


V1  V2     V3  V4  Class
0   2  50  12500  98      2
1   0  13   3250  28      2
2   1  16   4000  35      2
3   2  20   5000  45      2
4   1  24   6000  77      1

######  <p style="background-color:green;color:white;font-size:1.3em"> R2: Temos 748 instâncias </p> 

### 3 - Sobre a base de clusterização, marque a alternativa CORRETA:

In [15]:
print(wine.shape)
classes = wine['class'].unique()
size = len(classes)
print('Quantidade de classes: ', size)
print('Classes: ',  classes)

(178, 14)
Quantidade de classes:  3
Classes:  [1 2 3]


######  <p style="background-color:green;color:white;font-size:1.3em"> R3: A base tem 178 instâncias divididas em 3 classes. </p> 

### 4 - Sobre dados faltantes, marque a alternativa CORRETA:

In [16]:
blood.isnull().sum()

V1       0
V2       0
V3       0
V4       0
Class    0
dtype: int64

In [17]:
diabetes.isnull().sum()

age          0
deficit      0
c_peptide    0
dtype: int64

In [18]:
wine.isnull().sum()

class                             0
Alcohol                           0
Malic_acid                        0
Ash                               0
Alcalinity_of_ash                 0
Magnesium                         0
Total_phenols                     0
Flavanoids                        0
Nonflavanoid_phenols              0
Proanthocyanins                   0
Color_intensity                   0
Hue                               0
OD280%2FOD315_of_diluted_wines    0
Proline                           0
dtype: int64

######  <p style="background-color:green;color:white;font-size:1.3em"> R4: Nenhuma base tem dados faltantes. </p> 

### Quetões 5, 6 e 7

### 5 - Em relação a modelagem utilizando a regressão linear, marque a alternativa CORRETA sobre a métrica r2:

### 6 - Em relação a modelagem utilizando a regressão linear, marque a alternativa CORRETA sobre a métrica MAE:

### 7 - Em relação a modelagem utilizando a regressão linear, marque a alternativa CORRETA sobre a métrica MSE:

In [22]:
base = diabetes

X = base[['age', 'deficit']]
y = base[['c_peptide']]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

Avaliar as métricas R2, MAE e MSE.

In [23]:
# Regressão Linear
# Cria o objeto de classificação através do 
reg = LinearRegression()

# Realiza o treinamento do classificador
reg.fit(X_train, np.ravel(y_train))

#Realiza a previsão de classificaçao 
reg_pred = reg.predict(X_test)

#Métricas
R2 = metrics.r2_score(y_test, reg_pred)
MAE = mean_absolute_error(y_test, reg_pred)
MSE = mean_squared_error(y_test, reg_pred)

print('R2:', R2)
print('MAE:', MAE)
print('MSE:', MSE)

R2: 0.21880597788432776
MAE: 0.6107422323514552
MSE: 0.4745269473181977


######  <p style="background-color:green;color:white;font-size:1.3em"> R5:  O r2 encontrado é aproximadamente 0.22. </p> 
######  <p style="background-color:green;color:white;font-size:1.3em"> R6:  O MAE encontrado é aproximadamente 0.61. </p> 
######  <p style="background-color:green;color:white;font-size:1.3em"> R7:  O MSE encontrado é aproximadamente 0.47. </p> 

### Quetões 8, 9, 10 e 11

### 8 - Em relação a modelagem utilizando o SVM, marque a alternativa CORRETA sobre a métrica acurácia:

### 9 - Em relação a modelagem utilizando o SVM, marque a alternativa CORRETA sobre as métricas precision e recall:

### 10 - Em relação a modelagem utilizando o SVM, marque a alternativa CORRETA sobre a métrica f1:

### 11 - Em relação a modelagem utilizando o SVM, marque a alternativa CORRETA sobre a métrica AUROC:

In [28]:
base = blood

X = base[['V1', 'V2', 'V3', 'V4']]
y = base[['Class']]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

Acurácia, Precision, Recall, F1 e AUROC

In [31]:
# Cria o objeto de classificação através do 
clf_svm =  SVC() #gamma='auto',kernel='rbf', random_state=1

# Realiza o treinamento do classificador
clf_svm.fit(X_train, np.ravel(y_train))

#Realiza a previsão de classificaçao 
clf_svm_pred = clf_svm.predict(X_test)

#Métricas
accuracy = accuracy_score(y_test, clf_svm_pred)
precision = precision_score(y_test, clf_svm_pred)
recall = recall_score(y_test, clf_svm_pred)
F1 = f1_score(y_test, clf_svm_pred)
AUROC = roc_auc_score(y_test, clf_svm_pred)

print('Accuracy:', accuracy)
print('Precision:', precision)
print('Recall:', recall)
print('F1:', F1)
print('AUROC:', AUROC)

print('\nMétricas de Classificação\n', metrics.classification_report(y_test, clf_svm_pred))

Accuracy: 0.7486631016042781
Precision: 0.7473118279569892
Recall: 1.0
F1: 0.8553846153846154
AUROC: 0.5104166666666666

Métricas de Classificação
               precision    recall  f1-score   support

           1       0.75      1.00      0.86       139
           2       1.00      0.02      0.04        48

    accuracy                           0.75       187
   macro avg       0.87      0.51      0.45       187
weighted avg       0.81      0.75      0.65       187



######  <p style="background-color:green;color:white;font-size:1.3em"> R8: A acurácia encontrada é aproximadamente 0.75. </p> 
######  <p style="background-color:green;color:white;font-size:1.3em"> R9: Os valores de precision e recall encontrados são aproximadamente 1.0 e 0.02, respectivamente. </p> 
######  <p style="background-color:green;color:white;font-size:1.3em"> R10: A f1 encontrada é aproximadamente 0.04. </p> 
######  <p style="background-color:green;color:white;font-size:1.3em"> R11: A AUROC encontrada é aproximadamente 0.51, indicando que o modelo apresenta um resultado próximo de uma baseline aleatório. </p> 

### 12 - Em relação a modelagem utilizando o Kmeans, marque a alternativa CORRETA sobre o número de clusters

In [32]:
base = wine

#X = base.iloc[:,1:].columns
X = base[['Alcohol', 'Malic_acid', 'Ash', 'Alcalinity_of_ash',
       'Magnesium', 'Total_phenols', 'Flavanoids', 'Nonflavanoid_phenols',
       'Proanthocyanins', 'Color_intensity', 'Hue',
       'OD280%2FOD315_of_diluted_wines', 'Proline']]
y = base[['class']]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

In [33]:
def elbow_plotly(data, start_K, end_K, step):
    '''
    Generate an elbow plot to find optimal number of clusters
    graphing K values from start_K to end_K every step value
    '''
    
    score_list = []

    for i in range(start_K, end_K, step):
        print('K: ', i)
        kmeans = KMeans(i)
        model = kmeans.fit(data)
        score = model.score(data)
        score_list.append(abs(score))

    vals = go.Scatter(x =list(range(start_K, end_K, step)), y= score_list,
                            mode='lines+markers',
                            name='',
                            line=dict(color='blue', width=1))
    data = [vals]
    layout = go.Layout(title='Elbow Method for optimal K',
                       yaxis={'title': 'Average distance to centroid'},
                      xaxis={'title': 'K'})

    # Creating Figure
    fig = go.Figure(data=data, layout=layout)
    #plt.savefig('elbow_plot.png')
    # Showing figure/graph
    iplot(fig)

In [34]:
# Choosing number of k
print('Begin elbow_plotly function ...')
begin = time.time()
elbow_plotly(X_train, 1, 10, 1)
print('\nDuration:', time.time() - begin)

Begin elbow_plotly function ...
K:  1
K:  2
K:  3
K:  4
K:  5
K:  6
K:  7
K:  8
K:  9



Duration: 0.6270730495452881


######  <p style="background-color:green;color:white;font-size:1.3em"> R12: 3 clusters. Como já sabemos a quantidade de classes existentes, acredito que a resposta seja essa. O que se confirma utilizando o análise gráfica (cotovelo). </p> 

### Quetões 13, 14 e 15

### 13 - Em relação a modelagem utilizando o Kmeans, marque a alternativa CORRETA sobre a métrica Coeficiente de Silhueta:

### 14 - Em relação a modelagem utilizando o Kmeans, marque a alternativa CORRETA sobre a métrica Davies-Bouldin Score:

### 15 - Em relação a modelagem utilizando o Kmeans, marque a alternativa CORRETA sobre a métrica Mutual information:

Avaliar as métricas Coeficiente de Silhueta, Davies-Bouldin Score e Mutual
Information

In [85]:
# Cria o objeto de classificação através do 
kmeans = KMeans(n_clusters = 3) #, random_state = 42)

# Realiza o treinamento do classificador
kmeans.fit(X_train) #, np.ravel(y_train))

#Realiza a previsão de classificaçao 
kmeans_pred = kmeans.predict(X_test)

#Métricas
sil = silhouette_score(X_test, kmeans_pred)
davies = davies_bouldin_score(X_test, kmeans_pred)
mutual = mutual_info_score(y_test['class'].to_numpy(), kmeans_pred)

print('silhouette:', sil)
print('davies:', davies)
print('mutual_info_score:', mutual)

silhouette: 0.5519241838976598
davies: 0.5415115100039726
mutual_info_score: 0.526053317723497


######  <p style="background-color:green;color:white;font-size:1.3em"> R13: O coeficiente de silhueta encontrado é aproximadamente 0.55. </p> 
######  <p style="background-color:green;color:white;font-size:1.3em"> R14: O Davies-Bouldin Score encontrado é aproximadamente 0.54. </p> 
######  <p style="background-color:green;color:white;font-size:1.3em"> R14: O Mutual information encontrado é aproximadamente 0.53. </p> 